Domains: Consolidate TLD filters into single bar

jsnmoon opened the pull request:
This change consolidates TLD filters into one bar in the domain search experience, like so:

<img width="979" alt="wordpress_com" src="https://user-images.githubusercontent.com/4044428/39455184-67a43582-4c9c-11e8-812d-17ccc279d968.png">

This change enables the TLD filters bar in `development`, `horizon`, and `wpcalypso`.

# Testing instructions
1. Spin up this branch locally.
2. Navigate to `/start/domains` and enter a query.
3. Ensure that the TLD filters bar appears.
4. Ensure that clicking a TLD filter button works as expected.

a8ci18n commented on client/components/domains/search-filters/tld-filter-bar.jsx:
```diff
@@ -0,0 +1,205 @@
+					ref={ this.bindButton }
+					key="popover-button"
+				>
+					{ translate( 'More Extensions', {
```
Hi! I've found a possible matching string that has already been translated **1** time:
`translate( 'Extensions', { context: 'Refers to top level domain name extension, such as ".com"'} )` *ES Score: 19*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

jsnmoon replied:
Fixed. Now uses both `context` and `comment`.

delputnam commented:
@jsnmoon I wonder if perhaps we shouldn't show domains already listed to the left of the "more extensions" button in the popover since the initial list of TLDs available in the popover is exactly the same as the ones already shown in the bar.

jsnmoon commented:
@delputnam: To clarify, are you referring to this?

<img width="902" alt="wordpress_com" src="https://user-images.githubusercontent.com/4044428/39652720-bca0ed66-4fab-11e8-96b1-23db292715a2.png">

jsnmoon commented:
Fixed a handful of bugs, refactored component methods, removed custom sort scoring, and excluded suggested TLDs from the drop-down TLD token field.

I recommend moving custom TLD sort to the backend for better performance; I'll log this in our internal Trello.

klimeryk commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -949,6 +938,10 @@ class RegisterDomainStep extends React.Component {
 
+		const showTldFilterBar =
+			config.isEnabled( 'domains/kracken-ui/tld-filters' ) ||
+			abtest( 'showDomainTldFilterBar' ) === 'show';
```
If we have the `showDomainTldFilterBar` A/B test, do we really need the feature flag? It could be a bit confusing if we merge this, but don't enable the flag on staging and/or production. Especially since the feature flag is checked inside the component as well (but not the A/B test variation). So it ends up confusing - you force the `hide` variation, but the TLD bar is forced on production and wpcalypso. I'd keep it simple and just leave the A/B test one, unless I missed something.

jsnmoon replied:
Sounds good to me.

jsnmoon replied:
I actually went ahead and removed the A/B test from this PR. I'll be sure to spin up an A/B test for all the feature changes we're anticipating for the next redesign milestone.

klimeryk commented on client/components/domains/search-filters/tld-filter-bar.jsx:
```diff
@@ -0,0 +1,206 @@
+				position="bottom left"
+			>
+				<FormFieldset className="search-filters__token-field-fieldset">
+					<TokenField
```
Seems this needs a `maxSuggestions={ 500 }` or something, otherwise it defaults to `100`, which is probably not what we need/want (since it cuts of the TLD list quite early).

jsnmoon replied:
👍

a8ci18n commented on client/components/domains/search-filters/tld-filter-bar.jsx:
```diff
@@ -0,0 +1,207 @@
+				ref={ this.bindButton }
+				key="popover-button"
+			>
+				{ this.props.translate( 'More Extensions', {
```
Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'Extensions', { context: 'Refers to top level domain name extension, such as ".com"'} )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

jsnmoon commented:
Rebased from master and removed the A/B test setup from the PR. (A/B testing will be handled in a future PR that will cover more than just this feature.)

klimeryk commented on client/components/domains/register-domain-step/analytics.js:
```diff
@@ -95,3 +95,21 @@ export function recordShowMoreResults( searchQuery, pageNumber, section ) {
+		recordTracksEvent( 'calypso_domain_search_results_filter_reset', {
+			keys_to_reset: keysToReset,
+			section,
+			...filters,
```
This causes an error in the console log:
```
Tracks: Event `calypso_domain_search_results_filter_submit` will be rejected because property name `showExactMatchesOnly` does not match /^[a-z_][a-z0-9_]*$/. Please use a compliant property name.
```
(and same for other filters: `maxCharacters` and `includeDashes`)

Also:
```
Tracks: Unable to record event "calypso_domain_search_results_filter_submit" because nested properties are not supported by Tracks. Check 'tlds' on {...}
```

jsnmoon replied:
Ah, yes. You're completely right. Luckily, this has already been fixed in #24718 (which also added analytics tracking to filter resets and submissions). In #24718, the filter keys are snake-cased to resolve this issue.

Will rebase soon.

klimeryk commented on client/components/domains/search-filters/tld-filter-bar.jsx:
```diff
@@ -0,0 +1,194 @@
+		this.togglePopover();
+		this.props.onSubmit();
+	};
+	handleTokenChange = newTlds => {
```
Looks like something broke here (or in the component), because I can't seem to remove the tlds anymore from the `TokenField`.

To reproduce: just add any TLD (using the buttons or token field) and try to remove them using `X`. The `Reset` button works, though.

jsnmoon replied:
I can reproduce. My guess is that I must have messed up a rebase somewhere; fix incoming.

aidvu commented:
@jsnmoon so the fix with displaying the results is also in master? Rebase please, and I can test this one again.

jsnmoon commented:
> There's a certain breakpoint where the More extensions button goes under the rest (which is good!), but it feels like there should be some top/bottom margin there to separate them (@klimeryk)

Agreed - I could've sworn I added styling to address this exact issue. I'll verify and circle back.

> Looks like the bit that made the font smaller as suggestions' length increased got broken or removed along the way? Is this expected? (@klimeryk)

> Suggested domains: `jbyy798u0iokajkbiuh89iasdf.com` and `jbyy798u0iokajkbiuh89iasdf.blog`

The handling for this is [here](https://github.com/Automattic/wp-calypso/blob/master/client/components/domains/featured-domain-suggestions/index.jsx#L50). I'm not sure what's happening to cause featured suggestions stacking like your screenshot; when I reproduce using your query, it looks as expected: 

<img width="999" alt="wordpress_com" src="https://user-images.githubusercontent.com/4044428/39896100-0d0c3de2-546a-11e8-9950-9bd787a8d121.png">

I'll rebase this PR soon to the latest code. Could you give it a try once the rebase is complete? I'll be sure to ping you.

klimeryk commented on client/components/domains/register-domain-step/index.jsx:
```diff
@@ -170,6 +170,8 @@ class RegisterDomainStep extends React.Component {
 		onAddDomain: PropTypes.func,
 		onAddTransfer: PropTypes.func,
 		designType: PropTypes.string,
+		recordFilterSubmit: PropTypes.func.isRequired,
```
These two props are throwing warnings in the console:
```
Warning: Failed prop type: The prop `recordFilterSubmit` is marked as required in `RegisterDomainStep`, but its value is `undefined`.
```

To reproduce - go to Domain Search in Domain Management.

Additionally, looks like they are not used anywhere anymore?

jsnmoon replied:
Ah, I need to update those two props to be `recordFiltersSubmit` and `recordFiltersReset` instead.

klimeryk commented:
Other than that ^, everything looks great and works as expected.

klimeryk commented:
Sorry missed this part:
> @klimeryk: I rebased from master and reduced the number of suggested TLDs per CSS breakpoint. Could you let me know if you still run into the More Extensions button overflowing? I'm personally having some difficulty reproducing it.

Yup, it looks good now!

jsnmoon commented:
@klimeryk I just pushed out a change that better handles overflowing buttons like so. What do you think?

<img width="570" alt="domain_search_ _testing_ _wordpress_com" src="https://user-images.githubusercontent.com/4044428/39939243-d5943c44-5512-11e8-9f6a-4e326870ed65.png">

jsnmoon commented:
@sixhours Unfortunately, that's rather difficult due to how current flexbox specs are defined; there's no easy way to target the last item in an overflowing flex row. It's possible to manually target an element by index (ex. find the 5th TLD button and remove its right margin), but such implementation would be very brittle.

For example, we intend to dynamically suggest TLDs in the filter bar depending on our user's query in the future. It's possible that some TLD buttons will overflow onto the second row due to their length (ex. `.accountant`). Targetting styles via index would fail completely in this scenario.
